### PR TITLE
Sample config file generation

### LINKED ISSUE #34: Sample file generation
Add a mechanism for generating sample files from configuration. The process of using the software should be easier and less error-prone for users. Support both Istio and native Envoy.

### PR DESCRIPTION
**Pull request by rockspore**
It works as the following:
`./cli samples create -c config.yaml [--out-dir samples] [--native --host target-service-host --name target-service-cluster-name --prefix target-service-prefix --tls tls-directory]`

`--out-dir` defaults to `./samples` and the cli overwrites existing files.
Those target service related flags as well as tls are only for `--native`. Users are responsible for preparing the deployment related files for their target services if it's in Istio.

Open to suggestions.

To fix #34

### CODE REVIEW

**Comment by theganyo** [comment]
This looks very cool, I'll try it out. If we are eliminating the need for users to download the -envoy release package just to get the samples, that would be fantastic. 

Thoughts on the flags:

We might as well add short flags as well. 
Should we require target name and host instead of defaulting?
Maybe `out-dir` could just be `-o/--output`? 
It would be safer to default to failing if the target directory, and add a param like `-f/--force` to overwrite.

Thoughts?

**Comment by rockspore** [comment]
> This looks very cool, I'll try it out. If we are eliminating the need for users to download the -envoy release package just to get the samples, that would be fantastic.
> 
> Thoughts on the flags:
> 
> We might as well add short flags as well.
> Should we require target name and host instead of defaulting?
> Maybe `out-dir` could just be `-o/--output`?
> It would be safer to default to failing if the target directory, and add a param like `-f/--force` to overwrite.
> 
> Thoughts?

Short flags sounds good! But since the command is taking advantage of the shared.Resolve() right now, many characters such as `-o` and `-f` have been taken. If it makes sense to you, the change I can make is to stop using the Resolve() function and write another one to force the usage of a config yaml. As a result, using flags such as `-o/--organization` and `-e/--environment` would not work anymore.

I fully agree on the no-overwrite by default strategy. In fact the program didn't allow overwriting till I started the unit tests and went for the convenient route...Will address it.

**Comment by rockspore** [comment]
If the goal is to eliminate the need for the samples in -envoy repo, we may as well generate a `httpbin.yaml` file with its content commented out in case someone is just following the tutorial.

**Comment by theganyo** [comment]
Yes, removing the need for Resolve() sounds fine to me.

Re: httpbin, yes... we could do that as well. If a user could do the sample create, apply, and run, without any additional work... that would make it less error-prone. But it only makes sense if they're actually specifying httpbin as their target. I suppose we could make that the default (I know, reversing what I said on making those entries required) and generate the whole thing. Then, if they do specify a target, we don't create it.

**Comment by rockspore** [comment]
`-o`, `-h` are global flags, specially `-h` is taken by cobra for --help. So the shorthands for `--out` (changed from `--output`) and `--host` are not enabled. The rest of what we discussed has been implemented.

**Comment by rockspore** [comment]
Working to bin the template files into templates.go.

File: cmd/samples/samples.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+the httpbin example will be generated. Otherwise, users are responsible for preparing
+files related to deployment of their target services.`,
+		Args: cobra.NoArgs,
+
+		RunE: func(cmd *cobra.Command, _ []string) error {
+			err := s.loadConfig()
+			if err != nil {
+				return errors.Wrap(err, "loading config yaml file")
+			}
+			return errors.Wrap(s.createSampleConfigs(printf), "creating sample config files")
```
Comment (theganyo) [inline cmd/samples/samples.go:103]: We should emit a final message when it finishes successfully. Maybe remind them of the next steps: label the default namespace and run kubectl apply on the directory?

File: cmd/samples/samples.go [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+		},
+	}
+
+	c.Flags().StringVarP(&s.ConfigPath, "config", "c", "", "Path to Apigee Remote Service config file")
+	c.Flags().BoolVarP(&s.isNative, "native", "", false, "generate config for native envoy (otherwise assuming istio)")
+	c.Flags().BoolVarP(&s.overwrite, "force", "f", false, "force overwriting existing directory")
+	c.Flags().StringVarP(&s.outDir, "out", "", "./samples", "directory to create config files within")
+	c.Flags().StringVarP(&s.TargetService.Name, "name", "n", "httpbin", "target service name")
+	c.Flags().StringVarP(&s.TargetService.Host, "host", "", "httpbin.org", "target service host")
+	c.Flags().StringVarP(&s.TargetService.Prefix, "prefix", "p", "/", "target service prefix")
```
[the remark this answers is not in the document]
Comment (rockspore) [reply cmd/samples/samples.go:113]: Sounds good to me.

File: cmd/samples/samples.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+			err := s.loadConfig()
+			if err != nil {
+				return errors.Wrap(err, "loading config yaml file")
+			}
+			return errors.Wrap(s.createSampleConfigs(printf), "creating sample config files")
+		},
+	}
+
+	c.Flags().StringVarP(&s.ConfigPath, "config", "c", "", "Path to Apigee Remote Service config file")
+	c.Flags().BoolVarP(&s.isNative, "native", "", false, "generate config for native envoy (otherwise assuming istio)")
```
[the remark this answers is not in the document]
Comment (rockspore) [reply cmd/samples/samples.go:108]: Makes sense.

File: cmd/samples/samples.go (hunk 1 above)
Comment (rockspore) [reply cmd/samples/samples.go:103]: Sure.

File: cmd/samples/samples.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+		RunE: func(cmd *cobra.Command, _ []string) error {
+			err := s.loadConfig()
+			if err != nil {
+				return errors.Wrap(err, "loading config yaml file")
+			}
+			return errors.Wrap(s.createSampleConfigs(printf), "creating sample config files")
+		},
+	}
+
+	c.Flags().StringVarP(&s.ConfigPath, "config", "c", "", "Path to Apigee Remote Service config file")
```
[the remark this answers is not in the document]
Comment (rockspore) [reply cmd/samples/samples.go:107]: Actually, the config file description in the root.go is identical to this. I am going to change both to lower case.

File: cmd/samples/samples.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (rockspore) [reply cmd/samples/samples.go:113]: Sorry, but I already removed the prefix rewriting. This flag is taking the path for matching, i.e., for the Envoy router to take the request to the desired upstream cluster. Should we remove that as well?

[the remark this answers is not in the document]
Comment (theganyo) [reply cmd/samples/samples.go:113]: Gotcha. Hmm. But the only benefit would be to allow other endpoints to not be captured by the matcher, correct? Is that important for sample generation?

**Pull request merged**